src/proof_rules/calculus/soundness_checker.rs

```diff
@@ -0,0 +1,261 @@
+    pub is_loop: bool,
+    pub kind: ApproximationKind,
+}
+pub type ApproximationList = Vec<ApproximationListEntry>;
+
+// pub type SoundnessBlame = (Soundness, ApproximationList);
+
+#[derive(Debug, Clone, Default)]
+pub struct SoundnessBlame {
+    pub soundness: Soundness,
+    pub approximations: ApproximationList,
+    pub calculus: Option<Calculus>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum ApproximationKind {
+    #[default]
+    Exact,
+    Under,
+    Over,
+    Unknown,
+}
+
+impl ApproximationKind {
+    pub fn supremum(self, other: ApproximationKind) -> ApproximationKind {
+        match self.partial_cmp(&other) {
+            Some(std::cmp::Ordering::Equal) => self,
+            Some(std::cmp::Ordering::Less) => other,
+            Some(std::cmp::Ordering::Greater) => self,
+            None => ApproximationKind::Exact, // If they are incomparable, return Exact (which is the top element)
+        }
+    }
+    pub fn infimum(self, other: ApproximationKind) -> ApproximationKind {
+        match self.partial_cmp(&other) {
+            Some(std::cmp::Ordering::Equal) => self,
+            Some(std::cmp::Ordering::Less) => self,
+            Some(std::cmp::Ordering::Greater) => other,
+            None => ApproximationKind::Unknown, // If they are incomparable, return Unknown (which is the bottom element)
+        }
+    }
+}
+
+impl PartialOrd for ApproximationKind {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        match (self, other) {
+            (ApproximationKind::Exact, ApproximationKind::Exact) => Some(std::cmp::Ordering::Equal),
+            (ApproximationKind::Under, ApproximationKind::Under) => Some(std::cmp::Ordering::Equal),
+            (ApproximationKind::Over, ApproximationKind::Over) => Some(std::cmp::Ordering::Equal),
+            (ApproximationKind::Unknown, ApproximationKind::Unknown) => {
+                Some(std::cmp::Ordering::Equal)
+            }
+            (_, ApproximationKind::Exact) => Some(std::cmp::Ordering::Less),
+            (ApproximationKind::Exact, _) => Some(std::cmp::Ordering::Greater),
+            (ApproximationKind::Unknown, _) => Some(std::cmp::Ordering::Less),
+            (_, ApproximationKind::Unknown) => Some(std::cmp::Ordering::Greater),
+            (ApproximationKind::Under, ApproximationKind::Over)
+            | (ApproximationKind::Over, ApproximationKind::Under) => None,
+        }
+    }
+}
+
+impl fmt::Display for ApproximationKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ApproximationKind::Exact => write!(f, "exact approximation"),
+            ApproximationKind::Under => write!(f, "under-approximation"),
+            ApproximationKind::Over => write!(f, "over-approximation"),
+            ApproximationKind::Unknown => write!(f, "unknown approximation"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum Soundness {
```

Comment:
I am now just realizing: wouldn't it be much more direct to implement this type as a `struct` with two fields of type `bool`: `proofs` and `refutations`? Each of those states exactly corresponds to the items in the enum here. 

But the bools make more sense because it's more common to simply want to check whether the current thing (proof/refutation) would be sound, instead of matching all cases, no?

In addition, the `partial_cmp`/`infimum`/`supremum` impls would become much simpler. I think the *automatically derived* `PartialOrd`would even do the right thing! Infimum and supremum would correspond to field-wise `&&` resp. `||`.